### DISCUSSION TITLE
Relationships

### DISCUSSION THREAD

**Discussion by Jaycyn** [netreconlab/Parse-Swift#185] [General]
As a general question we would like to ensure we're implementing one->many relationship properly. We come from using Realm where a person with two dogs would look like this

```
let p = PersonClass(personName: "Jay)
let d0 = DogClass(dogName: "Spot")
let d1 = DogClass(dogName: "Rover")
p.dogList.add(objectsIn : [d0, 1])
try realm.write {
   realm.add(p)
}
```

The above code would create a person, both dogs and generate a  forward relationship from the person to each dog.

In Parse, to accomplish the same relationship this appears to be the process:

```
let p = PersonStruct(personName: "Jay")
let d0 = DogStruct(dogName: "Scraps")
let d1 = DogStruct(dogName: "Fluffy")

p.save { personResults in
    switch personResults {
    case .success(let savedPerson):
        [d0, d1].saveAll { dogResults in
            switch dogResults {
            case .success(let savedDogs):
                do {
                    let relation = savedPerson.relation
                    let dogs = savedDogs.compactMap { try? $0.get() }
                    guard let newRelations = try relation?.add("dogs", objects: dogs) else {
                        print("error adding relationship to dogs")
                        return
                    }

                    newRelations.save { relationResult in
                        switch relationResult {
                        case .success(let savedRelation):
                            print("saving relation succeeded")
                        case .failure(let errorRelation):
                            print("error saving relation: \(errorRelation.localizedDescription)")
                        }
                    }
                } catch {
                    print(error.localizedDescription)
                }
            case .failure(let errorDogs):
                print("error saving dogs: \(errorDogs.localizedDescription)")
            }
        }
    case .failure(let errorPerson):
        print("error saving person: \(errorPerson.localizedDescription)")
    }
}
```

We are looking for confirmation that is the correct process - and/or if there is any way to condense or shorten the code.

**Comment by cbaker6** [comment]
From the looks of your code, your implementation should work. There are at least 2 ways to create relationships in Parse; [Pointers](https://github.com/netreconlab/Parse-Swift/tree/main/ParseSwift.playground/Pages/8%20-%20Pointers.xcplaygroundpage) and [Relations](https://github.com/netreconlab/Parse-Swift/blob/main/ParseSwift.playground/Pages/12%20-%20Roles%20and%20Relations.xcplaygroundpage/Contents.swift). Both are demonstrated working implementations in the Playgrounds. You can reduce your code significantly and make it more readable by using the async/await methods available in the SDK

**Reply by cbaker6** [reply]
There's also information about the various data types in the server documentation https://docs.parseplatform.org/rest/guide/#data-types along a number of conversations online.

**Comment by cbaker6** [comment]
@Jaycyn I’ve seen your comments on the parseplatform discord. If you have additional questions about Parse-Swift, feel free to ask by opening up additional discussions or issues in this repo. Of course, for items related to Parse-Server or other SDKs, the parseplatform discord is the best place for that.

To see a feature comparison with this SDK against the rest of the parseplatform SDKs, you can visit here: https://github.com/netreconlab/Parse-Swift/discussions/72

As for async/await, this SDK fully supports it and the [API is fully documented](https://swiftpackageindex.com/netreconlab/Parse-Swift/documentation) and a ton of working examples in [Swift Playgrounds](https://github.com/netreconlab/Parse-Swift/tree/main/ParseSwift.playground/Pages)

**Comment by Jaycyn** [comment]
Thanks @cbaker6

We agree that Parse-Swift(OG) is quite a bit ahead of [Parse-SDK-iOS-OSX](https://github.com/parse-community/Parse-SDK-iOS-OSX), and it's pretty impressive. We worked with [Back4App](https://www.back4app.com), which is a great back end platform and we were also able to spin up a local MongoDB Community server for local development so we're not banging on Back4App all the time.

The relationships (per the thread question) are working for us - and thanks for the direction. That was a huge help during evaluation.

Our holdup now is we need queryable, on-device data and our clients cannot spin up on-device servers. It seems we can do that with the [Parse-SDK-iOS-OSX](https://github.com/parse-community/Parse-SDK-iOS-OSX) but not with Parse-Swift(OG) - it would require another database like CoreData of SwiftData and then we have a different relationship setup, two sets of queries and double the codebase. But I digress.

Thanks again.

**Reply by cbaker6** [reply]
> We agree that Parse-Swift(OG) is quite a bit ahead of [Parse-SDK-iOS-OSX](https://github.com/parse-community/Parse-SDK-iOS-OSX), and it's pretty impressive. We worked with [Back4App](https://www.back4app.com), which is a great back end platform and we were also able to spin up a local MongoDB Community server for local development so we're not banging on Back4App all the time.
> 
> The relationships (per the thread question) are working for us - and thanks for the direction. That was a huge help during evaluation.

Great to hear!

> Our holdup now is we need queryable, on-device data and our clients cannot spin up on-device servers. It seems we can do that with the [Parse-SDK-iOS-OSX](https://github.com/parse-community/Parse-SDK-iOS-OSX) but not with Parse-Swift(OG) - it would require another database like CoreData of SwiftData and then we have a different relationship setup, two sets of queries and double the codebase. But I digress.

I see, hopefully the parse-platform Objective-C SDK is able to be tailored to suit your needs

**Reply by cbaker6** [reply]
> So posing the question: does the Parse Apple SDK support async/await, or is that relegated to the Parse-Swift branch?

@Jaycyn [Parse-SDK-iOS-OSX](https://github.com/parse-community/Parse-SDK-iOS-OSX) **DOES NOT** support async/await. It has its own design for asynchronous calls that came out long before Swift async/await using [Bolts](https://github.com/parse-community/Bolts-ObjC). The feature comparison I mentioned before in https://github.com/netreconlab/Parse-Swift/discussions/185#discussioncomment-11067742 specifies this. In your example code, it works sometimes because you are simply getting "lucky".

Though there have been some minor merged PRs to [Parse-SDK-iOS-OSX](https://github.com/parse-community/Parse-SDK-iOS-OSX) and [Bolts](https://github.com/parse-community/Bolts-ObjC) to allow them to build with the latest Xcode. There have been minimal additions since version [1.19.0](https://github.com/parse-community/Parse-SDK-iOS-OSX/releases/tag/1.19.0) (2020). Most PR's afterward to `Parse-SDK-iOS-OSX` have been patch and maintaince PR's with a small amount of feature additions (see for your self by looking at the [release notes](https://github.com/parse-community/Parse-SDK-iOS-OSX/releases), most significant changes are file name refactoring so it works with SPM). Both ParseSwiftOG and ParseSwift are much further ahead except local storage.

**Comment by Jaycyn** [comment]
Thank you for the insight, Corey.

> In your example code, it works sometimes because you are simply getting "lucky".

We had a feeling that was the case - just needed confirmation. There's so little activity on the forums we didn't think it would be addressed so glad you saw it.

ParseSwiftOG is way ahead in pretty much all aspects and would significantly shorten development time on our re-write because of the rich feature set, which has similarities to Realm. That local-first aspect is a requirement though, so it rules out Parse unfortunately. We considered crafting a bridge to a local store similar to ParseCareKit, but that in itself is a significant effort.
